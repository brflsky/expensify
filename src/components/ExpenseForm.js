import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();
class ExpenseForm extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      description: props.expense.description ? props.expense.description : '',
      notes: props.expense.notes ? props.expense.notes : '',
      amount: props.expense.amount ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense.createdAt ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onNotesChange = (e) => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt)
      this.setState(() => ({ createdAt }));
  }
  onCalenderFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  }
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'Please enter required values'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        notes: this.state.notes,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  }
  render() {
    return (
      <div>
        <h3>ExpensForm</h3>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            id="description"
            type="text"
            placeholder="Expense Title"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onCalenderFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add some notes here"
            onChange={this.onNotesChange}
            value={this.state.notes}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;