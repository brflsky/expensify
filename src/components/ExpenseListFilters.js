import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setSortAsc, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  };


  onDateChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }));
  }

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  }

  onSortAscDescChange = (e) => {
    if (e.target.value === 'true') {
      this.props.setSortAsc(true);
    } else {
      this.props.setSortAsc(false);
    }
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  render () {
    return (
      <div>
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
        <select onChange={this.onSortChange}
          value={this.props.filters.sortBy}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <select onChange={this.onSortAscDescChange}
          value={this.props.filters.sortAsc ? 'true' : 'false'}>
          <option value="true">Asc</option>
          <option value="false">Desc</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDateChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>

    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => {
  return {
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setSortAsc: (e) => dispatch(setSortAsc(e)),
    setTextFilter: (e) => dispatch(setTextFilter(e))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);