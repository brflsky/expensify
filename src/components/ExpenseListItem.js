import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { removeExpense } from '../actions/expenses';

export const ExpenseListItem = (props) => {
  // console.log('props', props);
  return (
    <div>
      <p>{props.index} <Link to={`/edit/${props.id}`} >{props.description}</Link>  {props.amount} {props.createdAt}

        <button onClick={() => { props.dispatch(removeExpense(props)); }}>X</button>
      </p>
    </div>
  );
};

export default connect()(ExpenseListItem);
