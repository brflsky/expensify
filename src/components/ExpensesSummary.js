import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import select from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
  <div className="page-header">
    <div className="content-container">
      <h1 className="page-header__title">
        Viewing <span>{props.expenses.length}</span>. Total cost: <span>{numeral(getExpensesTotal(props.expenses) / 100).format('$0,0.00')}</span>
      </h1>
      <div className="page-header__actions">
        <Link className="button" to="/create" >Add Expense</Link>
      </div>
    </div>
  </div>
);


const mapStateToProps = (state) => ({
  expenses: select(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);
