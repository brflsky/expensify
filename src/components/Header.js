import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { startLogout } from '../actions/auth';

export const Header = ({ logout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard" ><h1>Expensify</h1></Link>
        <button className="button button--link" onClick={logout}>Logout</button>
      </div>
    </div>
  </header>

);

const mapDispatchToProps = (dispatch) => ({
  logout: () => dispatch(startLogout())
});

export default withRouter(connect(null, mapDispatchToProps)(Header));
