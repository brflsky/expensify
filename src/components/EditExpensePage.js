import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    //console.log('updated',expense);
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  onClick = () => {
    this.props.removeExpense (this.props.expense);
    this.props.history.push('/');
  }
  render() {
    return  (
      <div>
        <ExpenseForm 
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onClick}
        >X</button>
      </div>
    );
  }
}


const mapStateToProps = (state, props) =>{
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};
const mapsDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id,expense)),
    removeExpense: expense => dispatch(removeExpense(expense))
  }
}

export default connect(mapStateToProps, mapsDispatchToProps)(EditExpensePage);  