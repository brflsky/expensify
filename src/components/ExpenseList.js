import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectedExpenseList from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
      {props.expenses.length > 0 ?  props.expenses.map((expense, index) => (
      <ExpenseListItem 
        key={expense.id}
        id={expense.id}
        index={index+1}
        description={expense.description}
        amount={expense.amount}
        createdAt={expense.createdAt}
      />
    )) : 
    <p>Nothing to show... ad something</p>
  }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectedExpenseList(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);

// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses
//   }
// })(ExpenseList);

// export default ConnectedExpenseList;