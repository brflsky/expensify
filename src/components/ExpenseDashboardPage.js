import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';



const ExpansifyDashboardPage = () => (
  <div>From Dashboard
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpansifyDashboardPage;