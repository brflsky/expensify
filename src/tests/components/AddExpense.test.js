import React from 'react';
import {shallow}  from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import {expenses} from '../fixtures';

let addExpense, history, wrapper;

beforeEach(() => {
  addExpense = jest.fn();
   history = {push: jest.fn()};
   wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history}/>);
});

test('should render Add Expense Page', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should correct submit form', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(addExpense).toHaveBeenLastCalledWith(expenses[1]);
});