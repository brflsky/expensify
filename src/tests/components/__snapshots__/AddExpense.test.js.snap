// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Add Expense Page 1`] = `
<div>
  <h1>
     Add Expense 
  </h1>
  <ExpenseForm
    expense={Object {}}
    onSubmit={[Function]}
  />
</div>
`;
